ethers-core/src/types/ens.rs

```diff
@@ -85,8 +72,57 @@ impl<'de> Deserialize<'de> for NameOrAddress {
         D: Deserializer<'de>,
     {
         let inner = Address::deserialize(deserializer)?;
+        Ok(Self::Address(inner))
+    }
+}
+
+impl From<&str> for NameOrAddress {
+    fn from(s: &str) -> Self {
+        Self::from_str(s).unwrap()
+    }
+}
+
+impl From<String> for NameOrAddress {
+    fn from(s: String) -> Self {
+        Self::Name(s)
+    }
+}
+
+impl From<&String> for NameOrAddress {
+    fn from(s: &String) -> Self {
+        Self::Name(s.clone())
+    }
+}
+
+impl From<Address> for NameOrAddress {
+    fn from(s: Address) -> Self {
+        Self::Address(s)
+    }
+}
 
-        Ok(NameOrAddress::Address(inner))
+impl FromStr for NameOrAddress {
+    type Err = Infallible;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Self::Name(s.to_string()))
+    }
+}
+
+impl NameOrAddress {
+    /// Maps Address(&a) to Some(a) by cloning the value and Name to None.
+    pub fn as_address(&self) -> Option<Address> {
+        match self {
+            Self::Address(a) => Some(*a),
+            Self::Name(_) => None,
+        }
+    }
+
+    /// Maps Name(&n) to Some(n) by cloning the value and Address to None.
+    pub fn as_name(&self) -> Option<String> {
```

Comment:
this should return `&str` so caller can decided whether to clone

Reply:
Ok, same for `as_address`?

Reply:
`Address` is `Copy` so nbd there, but for consistency maybe it makes sense